src/gamepad/mapping.vala

```diff
@@ -0,0 +1,236 @@
+// This file is part of GNOME Games. License: GPLv3
+
+private class LibGamepad.DPad : Object {
+	public InputType types[4];
+	public int values[4];
+	public int axis_values[2];
+
+	public DPad () {
+		axis_values[0] = axis_values[1] = 0;
+		types[0] = types[1] = types[2] = types[3] = InputType.INVALID;
+	}
+}
+
+public class LibGamepad.Mapping : Object {
+	private InputType[] buttons_type;
+	private int[] buttons_value;
+	private InputType[] axes_type;
+	private int[] axes_value;
+	private DPad[] dpads;
+
+	public Mapping.from_sdl_string (string? mapping_string) throws MappingError {
+		if (mapping_string == null || mapping_string == "")
+			throw new MappingError.NOT_A_MAPPING ("The mapping string can't be null nor empty.");
+
+		var mappings = mapping_string.split (",");
+		foreach (var mapping in mappings) {
+			var splitted_mapping = mapping.split (":");
+			if (splitted_mapping.length != 2)
+				continue;
+
+			var mapping_key = mapping.split (":")[0];
+			var mapping_value = mapping.split (":")[1];
+			var type = parse_input_type (mapping_key);
+			if (type == InputType.INVALID) {
+				if (mapping_key != "platform")
+					debug ("Invalid token : %s", mapping_key);
+
+				continue;
+			}
+			int parsed_value;
+			switch (type) {
+			case InputType.BUTTON:
+				parsed_value = parse_button (mapping_key);
+
+				break;
+			case InputType.AXIS:
+				parsed_value = parse_axis (mapping_key);
+
+				break;
+			default:
+				continue;
+			}
+			switch (mapping_value[0]) {
+			case 'h':
+				var dpad_parse_array = mapping_value[1:mapping_value.length].split (".");
+				var dpad_index = int.parse (dpad_parse_array[0]);
+				var dpad_position_2pow = int.parse (dpad_parse_array[1]);
+				int dpad_position = 0;
+				while (dpad_position_2pow > 1) {
+					dpad_position_2pow >>= 1;
+					dpad_position++;
+				}
+				while (dpads.length <= dpad_index)
+					dpads += new DPad ();
+				dpads[dpad_index].types[dpad_position] = type;
+				dpads[dpad_index].values[dpad_position] = parsed_value;
+
+				break;
+			case 'b':
+				var button = int.parse (mapping_value[1:mapping_value.length]);
+				while (buttons_type.length <= button)
+					buttons_type += InputType.INVALID;
+				if (buttons_value.length <= button)
+					buttons_value.resize (button + 1);
+				buttons_type[button] = type;
+				buttons_value[button] = parsed_value;
+
+				break;
+			case 'a':
+				var axis = int.parse (mapping_value[1:mapping_value.length]);
+				while (axes_type.length <= axis)
+					axes_type += InputType.INVALID;
+				if (axes_value.length <= axis)
+					axes_value.resize (axis + 1);
+				axes_type[axis] = type;
+				axes_value[axis] = parsed_value;
+
+				break;
+			}
+		}
+	}
+
+	public MappedEvent get_dpad_mapping (int dpad_index, int dpad_axis, int dpad_value) {
+		var event = MappedEvent ();
+		int dpad_position;
+		var dpad = dpads[dpad_index];
+		if (dpad_value == 0)
```

Comment:
This may be a good candidate to the ?: operator, maybe something like:
```
var ??? = dpad_value == 0 ? dpad.axis_values[dpad_axis] : dpad_value;
int dpad_position = (??? + dpad_axis + 4) % 4;
```

I'm not saying it's better, it's just an idea, do as you prefer, your way is fine too.